packages/alfa-dom/src/get-attribute.ts

```diff
@@ -17,36 +46,144 @@ const attributeMaps: WeakMap<Element, Map<string, string>> = new WeakMap();
-      const { prefix, localName, value } = attributes[i];
+  switch (selectorNamespace) {
+    case null:
+      attributeValue = getAttributeValue(element, name);
+      break;
+    case "*":
+      attributeValue = getWildcardAttributeValue(element, name);
+      break;
+    default:
+      attributeValue = getNamespaceAttributeValue(
+        element,
+        name,
+        selectorNamespace
+      );
+  }
 
-      const qualifiedName =
-        prefix === null ? localName : `${prefix}:${localName}`;
+  if (attributeValue === null) {
+    return null;
+  }
 
-      attributeMap.set(qualifiedName, value);
+  if (Array.isArray(attributeValue)) {
+    return attributeValue.map(value => applyOptions(value, options));
+  }
+
+  return applyOptions(attributeValue, options);
+}
+
+function getAttributeValue(
+  element: Element,
+  qualifiedName: string
+): string | Array<string> | null {
+  const { name, namespace } = splitQualifiedName(qualifiedName);
+  const attributes = getAttributeMap(element).get(name);
+
+  if (attributes === undefined) {
+    return null;
+  }
+
+  for (let i = 0, n = attributes.length; i < n; i++) {
+    const { prefix, value } = attributes[i];
+    if (prefix === namespace) {
+      return value;
     }
+  }
+
+  return null;
+}
+
+function getNamespaceAttributeValue(
+  element: Element,
+  name: string,
+  selectorNamespace: Namespace
+): string | null {
+  const attributes = getAttributeMap(element).get(name);
+
+  if (attributes === undefined) {
+    return null;
+  }
+
+  for (let i = 0, n = attributes.length; i < n; i++) {
+    const { value } = attributes[i];
+    const attributeNamespace = getAttributeNamespace(
+      attributes[i],
+      element,
+      element
+    );
 
-    attributeMaps.set(element, attributeMap);
+    if (attributeNamespace === selectorNamespace) {
+      return value;
+    }
   }
 
-  let value = attributeMap.get(name);
+  return null;
+}
 
-  if (value === undefined) {
+function getWildcardAttributeValue(
+  element: Element,
+  name: string
+): string | Array<string> | null {
+  const attributes = getAttributeMap(element).get(name);
+
+  if (attributes === undefined) {
     return null;
   }
 
+  const values: Array<string> = [];
+
+  for (let i = 0, n = attributes.length; i < n; i++) {
+    values.push(attributes[i].value);
+  }
+
+  return values;
+}
+
+function getAttributeMap(element: Element): Map<string, Array<Attribute>> {
+  let attributeMap = attributeMaps.get(element);
+
+  if (attributeMap !== undefined) {
+    return attributeMap;
+  }
+
+  attributeMap = new Map();
+
+  const { attributes } = element;
+
+  for (let i = 0, n = attributes.length; i < n; i++) {
+    const attribute = attributes[i];
+    const { localName } = attribute;
+
+    let attributeArray = attributeMap.get(localName);
+
+    if (attributeArray === undefined) {
+      attributeArray = [];
+    }
+
+    attributeArray.push(attribute);
+    attributeMap.set(localName, attributeArray);
```

Comment:
This line can be moved to just after line 174 to ensure that the map is only touched when initializing the array.